compiler-rt/lib/radsan/radsan.cpp

```diff
@@ -7,26 +7,137 @@
 */
 
 #include <radsan/radsan.h>
+
 #include <radsan/radsan_context.h>
+#include <radsan/radsan_flags.h>
 #include <radsan/radsan_interceptors.h>
-#include <unistd.h>
+
+#include "sanitizer_common/sanitizer_atomic.h"
+#include "sanitizer_common/sanitizer_common.h"
+#include "sanitizer_common/sanitizer_flag_parser.h"
+#include "sanitizer_common/sanitizer_flags.h"
+
+using namespace __sanitizer;
+
+namespace radsan {
+static __sanitizer::atomic_uint8_t radsan_inited{};
```

Comment:
Why do we need to use the `__sanitizer` atomic types here rather than just `std::atomic`?

Reply:
Good question. I think it is possibly so these can run on systems that don't have standard libraries.

This hunch (and it's 100% a hunch, could be misinformed) was based on something I saw in MSAN, and is also in tsan and hwasan

```bash
compiler-rt/lib/hwasan/hwasan.cpp
37:// ACHTUNG! No system header includes in this file.

compiler-rt/lib/tsan/rtl/tsan_rtl.h
19://   - No system headers included in header files (*).

compiler-rt/lib/msan/msan.cpp
34:// ACHTUNG! No system header includes in this file.
```

Reply:
After fussing with this a bit, there are very easy ways for you to segfault in these sanitizers. These internal types are defined to not use any system calls that may or may not invoke an intercepted function.


For instance, I was trying to make a Meyer's singleton of a big type. Because it was statically initialized, and was large enough, the compiler attempted to put a `pthread_mutex_lock` around the call. This happened in such a way that we were in a recursive loop (intercepting the lock, trying to initialize the member with a lock, intercepting the lock) and it segfaulted.

It seems as if you stay in the __sanitizer special types, it prevents you from shooting yourself in the foot.